obtain the grafana and tracing version from the internal in-cluster url

jmazzitelli opened the pull request:
fixes: https://github.com/kiali/kiali/issues/7758

jmazzitelli commented:
to test, install Istio and Grafana (just start up a minikube cluster and run `hack/istio/install_istio_via_istioctl.sh`).

Then build and push the dev images and install kiali (`make -e CLUSTER_TYPE=minikube build build-ui test cluster-push operator-create kiali-create`).

If you go to the Kiali UI Mesh page, and click the Grafana node, you will see "unknown" Grafana version - that's because it isn't configured yet.

Now edit the Kiali CR and set `external_services.grafana.in_cluster_url` to `http://grafana.istio-system:3000`. Set the url (`external_services.grafana.url`) to some bogus URL like `http://foo`. After you save the Kiali CR, wait for Kiali to restart and then go look at the Mesh page again in the UI and you should see the version now.

This shows that the version will be obtained by going over the in cluster URL. If you don't set that (but if you set the `url` to the valid endpoint) then it should still work -- the `url` is the fallback if `in_cluster_url` isn't set).

josunect reviewed (commented):
I was testing this in different scenarios, and I think it works as expected expect for the last one: 

in_cluster_url is set and valid, url incorrect. The version is set and there is access to Grafana: 
![image](https://github.com/user-attachments/assets/5e5fdc78-12fc-491b-849e-6c821bb481c9)

in_cluster_url is set and invalid, url is valid. The version is not set and there is no access to Grafana: 
![image](https://github.com/user-attachments/assets/679c453c-47e9-4f31-8746-920a87905d35)

in_cluster_url is not set, url is valid. The version is set and there is no access to Grafana: 
![image](https://github.com/user-attachments/assets/52645476-ab0a-4eb7-8507-b3ee819624d2)

in_cluster_url is set to '', and url is set to a valid URL. There is access to Grafana but there is no version in Grafana. Shouldn't this fall back to use the url? 
![image](https://github.com/user-attachments/assets/76d6dfc9-d71c-45f6-8219-1264cb3b9121)

Another question that I have is, should this behavior be the same for tracing and Prometheus urls?

jmazzitelli commented:
> Another question that I have is, should this behavior be the same for tracing and Prometheus urls?

Prometheus doesn't have an "in_cluster_url" - there's [only one URL setting](https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.external_services.prometheus.url)

Yes, should be the same as tracing. Does this bug exist for tracing too?

jmazzitelli commented:
> in_cluster_url is set to '', and url is set to a valid URL. There is access to Grafana but there is no version in Grafana. Shouldn't this fall back to use the url?

I just tried it.

Set it using your test scenario:
```
$ kubectl get kiali kiali -n kiali-operator -ojsonpath='{.spec.external_services}'|jq
```
```json
{
  "grafana": {
    "in_cluster_url": "",
    "url": "http://grafana.istio-system:3000"
  }
}
```

I see Grafana version in the Mesh page, and the api/status shows it:

```
$ curl -k http://localhost:20001/kiali/api
{
...
  "externalServices": [
...
    {
      "name": "Grafana",
      "version": "11.0.0",
      "url": "http://grafana.istio-system:3000",
      "tempoConfig": {}
    }
  ],
```

Mesh page shows this:

![image](https://github.com/user-attachments/assets/58b04211-cbf4-43aa-92cb-f859f71f6ce1)

jmazzitelli commented:
Looks like tracing has the same issue:
```
2024-09-18T14:13:40Z INF jaeger version check failed: url=[http://foo], code=[0]
```
That's with this:

```
$ kubectl get kiali kiali -n kiali-operator -ojsonpath='{.spec.external_services.tracing}'|jq
{
  "enabled": true,
  "in_cluster_url": "http://tracing.telemetry:16685/jaeger",
  "url": "http://foo",
  "use_grpc": true
}
```

jmazzitelli commented:
I created a new issue to include both tracing and grafana https://github.com/kiali/kiali/issues/7758
This PR will solve both issues. I need to update this PR to fix tracing, too.

josunect commented:
> > in_cluster_url is set to '', and url is set to a valid URL. There is access to Grafana but there is no version in Grafana. Shouldn't this fall back to use the url?
> 
> I just tried it.
> 
> Set it using your test scenario:
> 
> ```
> $ kubectl get kiali kiali -n kiali-operator -ojsonpath='{.spec.external_services}'|jq
> ```
> 
> ```json
> {
>   "grafana": {
>     "in_cluster_url": "",
>     "url": "http://grafana.istio-system:3000"
>   }
> }
> ```
> 
> I see Grafana version in the Mesh page, and the api/status shows it:
> 
> ```
> $ curl -k http://localhost:20001/kiali/api
> {
> ...
>   "externalServices": [
> ...
>     {
>       "name": "Grafana",
>       "version": "11.0.0",
>       "url": "http://grafana.istio-system:3000",
>       "tempoConfig": {}
>     }
>   ],
> ```
> 
> Mesh page shows this:
> 
> ![image](https://private-user-images.githubusercontent.com/2029470/368576590-58b04211-cbf4-43aa-92cb-f859f71f6ce1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY2NjgzNTMsIm5iZiI6MTcyNjY2ODA1MywicGF0aCI6Ii8yMDI5NDcwLzM2ODU3NjU5MC01OGIwNDIxMS1jYmY0LTQzYWEtOTJjYi1mODU5ZjcxZjZjZTEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkxOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MThUMTQwMDUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTgyMzdlYzI2MzNkOGFkZjZmNTE4MTFlMjViNjkyYTI1NmM0N2QyNTNjZjQxZDBjMTdiYTFhMTMzMTZhZmU5ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.dFZagJGB7eJ2d59fj2KCuOplIXIDrJPvEhq4-wKT5t0)

Sorry, @jmazzitelli , I meant to say the opposite (The screenshot is fine). `in_cluster_url` is valid, and `url` is set to "". Shouldn't this get the `in_cluster_url`? I understand it has precedence:

![image](https://github.com/user-attachments/assets/126b5f69-0044-4c69-a319-9a1f89b64362)

jmazzitelli commented:
> `in_cluster_url` is valid, and `url` is set to "". Shouldn't this get the `in_cluster_url`?

I think it is because of this code: https://github.com/kiali/kiali/blob/master/grafana/grafana.go#L210

This actually makes sense to me because in Kiali's Grafana integration (unlike our Tracing integration), you need an external public URL defined. Otherwise, it doesn't make sense. Because all the Grafana integration does is put external links in the Metrics pages for the user to click and go directly to the Grafana UI. So without this public `url` setting, the integration is basically a no-op -- it doesn't do anything.

This is different with our Tracing integration - we might not show the "Distributed Tracing" link in the left-hand nav, but we still provide tracing data directly within the Kiali UI (think of the bubble charts, or the purple highlights in the topology graph). So it is possible to not want to define a `url` for tracing, but still set `in_cluster_url`.

But without a `url` in Grafana, what does the Kiali UI provide?

Am I misunderstanding the Kiali Grafana integration? Do we actually have functionality in the Kiali UI where we simply query Grafana for data but do not need to show the user external links via that `url` ?

josunect commented:
> > `in_cluster_url` is valid, and `url` is set to "". Shouldn't this get the `in_cluster_url`?
> 
> I think it is because of this code: https://github.com/kiali/kiali/blob/master/grafana/grafana.go#L210
> 
> This actually makes sense to me because in Kiali's Grafana integration (unlike our Tracing integration), you need an external public URL defined. Otherwise, it doesn't make sense. Because all the Grafana integration does is put external links in the Metrics pages for the user to click and go directly to the Grafana UI. So without this public `url` setting, the integration is basically a no-op -- it doesn't do anything.
> 
> This is different with our Tracing integration - we might not show the "Distributed Tracing" link in the left-hand nav, but we still provide tracing data directly within the Kiali UI (think of the bubble charts, or the purple highlights in the topology graph). So it is possible to not want to define a `url` for tracing, but still set `in_cluster_url`.
> 
> But without a `url` in Grafana, what does the Kiali UI provide?
> 
> Am I misunderstanding the Kiali Grafana integration? Do we actually have functionality in the Kiali UI where we simply query Grafana for data but do not need to show the user external links via that `url` ?

@jmazzitelli yes, I think you are right about the Grafana and Kiali integration functionality. 

But, IMO, If checking the Grafana version will use the `in_cluster_url` first, I would expect this to be consistent and see the Grafana version if `in_cluster_url` is correct, independently of the `url` value. Because even if the version is "unknown", I don't think it warns the user that the `url` is wrong (I would expect to maybe see the Grafana component in red)

So, here, I don't see the Grafana version, and I see that Grafana is in green color, but there is an info message indicating that the links to Grafana are off:

![image](https://github.com/user-attachments/assets/5c16a0a3-76c4-4a0e-974a-89a0b4b0378d)

It could be difficult to know what is wrong?

jmazzitelli commented:
OK, @josunect ... check out the latest commit. That should fix that use case where "url" is empty string but in_cluster_url is set correctly.

jmazzitelli commented:
Latest commit now obtains tracing version over `in_cluster_url` (falls back to `url` if in_cluster_url is not specified).

Note that also we can't get version if GRPC is enabled... today, when GRPC is enabled, you will see "unknown" for version. This new code now attempts to get the version over HTTP, even if GRPC is enabled. It just does a best guess for the URL to use (it simply removes the port number from the URL, with the hopes the tracing server has an HTTP endpoint over the default port (80 for http, 443 for https) and that the HTTP API is even enabled in the tracing server).

josunect commented on status/versions.go:
```diff
@@ -80,16 +81,36 @@ func tracingVersion(conf *config.Config, homeClusterSAClient kubernetes.ClientIn
+		parsedUrl, err := url.Parse(versionUrl)
+		if err == nil {
+			// strip the port - if the URL is http, it'll go over 80, if https, it'll go over 443
+			if host := parsedUrl.Hostname(); host != "" {
```
I would strip the port just for Jaeger, because Tempo is going to set the URL for the http client with the port also when using grpc (And then it wouldn't make sense to remove the port, IMO). Also, the Tempo operator has an option to setup grpc in the same port as http, `stream_over_http_enabled`, so http and GRPC would be using the same port and the version would be available. 
For example, in this configuration: 
```
      tracing:
        enabled: true
        grpc_port: 9095
        in_cluster_url: http://tempo-cr-query-frontend.tempo:3200
        provider: tempo
        url: 'http://tempo-cr-query-frontend.tempo:3200'
        use_grpc: true
```
With this is not possible to get the version: 
![image](https://github.com/user-attachments/assets/9bf5eeb5-4175-4c37-9ffd-7c7803c3a673)

But the code that we have now makes it possible: 
![image](https://github.com/user-attachments/assets/e30f3753-4237-4f4f-acb4-e4ba6a260ef7)

jmazzitelli replied:
OK, check the latest commit. I just moved that block of code down into the "if-jaeger" section.

josunect reviewed (commented):
I have done some tests and it looks good, I've just added one comment for tracing

jmazzitelli commented:
> I have done some tests and it looks good, I've just added one comment for tracing

@josunect I modified the code based on that last comment you posted. See the latest commit.

josunect reviewed (approved):
LGTM!

![image](https://github.com/user-attachments/assets/99159626-2405-4805-bbf3-18d6d943d6ab)

![image](https://github.com/user-attachments/assets/a7dea306-bc4f-4f2a-b9fb-9d0ceb91c0b0)
